pontoon/static/js/widgets/table.js

```diff
@@ -0,0 +1,153 @@
+
+import Checkbox from 'widgets/checkbox';
+
+
+export class CheckboxTable extends React.Component {
+
+    constructor (props) {
+        super(props);
+        this.visible = [];
+        this.state = {checked: new Set()};
+        this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
+        this.handleSelectAll = this.handleSelectAll.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.handleTableChange = this.handleTableChange.bind(this);
+        this.handleTableSortChange = this.handleTableSortChange.bind(this);
+        this.handleTableResize = this.handleTableResize.bind(this);
+        this.renderCheckbox = this.renderCheckbox.bind(this);
+        this.renderSelectAllCheckbox = this.renderSelectAllCheckbox.bind(this);
+    }
+
+    get columnSelect () {
+        return {
+            Header: this.renderSelectAllCheckbox,
+            Cell: this.renderCheckbox,
+            sortable: false,
+            width: 45};
+    }
+
+    get defaultPageSize () {
+        return 5;
+    }
+
+    get pruned() {
+        // Returns a copy of the checked set with any resource paths that are
+        // not in `this.visible` removed
+        let {checked} = this.state;
+        return new Set([...checked].filter(v => (this.visible.indexOf(v) !== -1)))
+    }
+
+    get selected () {
+        // get the pruned list of selected resources, returns all=true/false
+        // if all visible resources are selected
+        // some rows can be empty strings if there are more visible rows than
+        // resources
+        const checked = this.pruned
+        const all = (
+            checked.size > 0
+                && checked.size === this.visible.filter(v => v !== '').length);
+        return {all, checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            this.visible.length = 0;
+            this.setState({checked: new Set()});
+        }
+    }
+
+    handleCheckboxClick (evt) {
+        // adds/removes paths for submission
+        let {checked} = this.state;
+        checked = new Set(checked);
+        evt.target.checked ? checked.add(evt.target.name) : checked.delete(evt.target.name);
+        this.setState({checked});
+    }
+
+    handleSelectAll () {
+        // user clicked the select all checkbox...
+        // if there are some resources checked already, all are removed
+        // otherwise all visible are checked.
+        let {checked} = this.state;
+        checked = checked.size > 0 ? new Set() : new Set([...this.visible.filter(x => x)]);
+        return this.setState({checked});
+    }
+
+    async handleSubmit (evt) {
+        // after emitting handleSubmit to parent with list of currently
+        // checked, clears the checkboxes
+        evt.preventDefault();
+        const {checked} = this.state;
+        await this.props.handleSubmit({data: [...checked]});
+        await this.setState({checked: new Set()});
```

Comment:
`setState()` doesn't return a `Promise` (in fact, it doesn't return anything), and `await` is not needed. All `return` statements for `this.setState({...})` are unnecessary too.